### PR TITLE
Add blockwise FP8 in roofline_utils and seperate benchmark script

### PR DESCRIPTION
**Pull request by iamzainhuda**
### Summary

Added FP8 blockwise linear benchmarking script. This will let us compare performance of our FP8 blockwise linear layer against FP16 linear layer across the DeepSeek V3 shapes.

Added
- A canonical DSV3 16B/671B FFN shape generator selected via
  `--shape_gen_name=dsv3-16b-671b`.
- Default DSV3 training-shape parameters: `M=seq_len=4096`, `dim=7168`,
  `inter_dim=18432`.
- Blockwise FP8 roofline estimates in `roofline_utils`, including FP8 GEMM time
  and modeled quantization overhead.
- GPU-name normalization in `roofline_utils` so device names with suffixes such
  as `NVIDIA H100 80GB HBM3` resolve to the existing roofline spec key.
- A focused `Float8BlockwiseLinear` benchmark script at
  `benchmarks/prototype/blockwise_fp8_training/bench_linear_roofline.py` which
  measures BF16 and blockwise FP8 fwd/bwd time, then reports measured speedup as
  a ratio of the shared roofline speedup target.
- README coverage for the focused benchmark output columns.

`blockwise_fp8_training` is intentionally roofline-only in this shared roofline script.
If `--do_benchmarks=True` is used with this recipe, the script raises an
unsupported assertion and asks the user to run roofline estimates with
`--do_benchmarks=False`.

### Results

```bash
python benchmarks/prototype/blockwise_fp8_training/bench_linear_roofline.py \
  --outfile=/tmp/blockwise_linear_roofline_dsv3_scaled_mm_rerun.csv \
  --iterations=20
```

Environment:

- GPU: NVIDIA H100 80GB HBM3
- PyTorch: `2.13.0a0+gitd129991`
- torchao: `0.17.0+gitcf0b50ae1`
- Shape set: `dsv3-16b-671b`
- Backend: `blockwise_scaled_mm`
- Compile: `False`

Roofline target:

- Roofline BF16 time: `4.209 ms`
- Roofline FP8 GEMM time: `2.103 ms`
- Roofline FP8 quantization overhead: `0.647 ms`
- Roofline FP8 total time: `2.750 ms`
- Roofline speedup estimate: `1.530x`

| layer | shape `(M,K,N)` | BF16 roofline ms | FP8 GEMM roofline ms | FP8 overhead ms | FP8 total ms | roofline speedup |
|---|---:|---:|---:|---:|---:|---:|
| `dsv3.ffn.w1` | `(4096,7168,18432)` | 4.209 | 2.103 | 0.647 | 2.750 | 1.530x |
| `dsv3.ffn.w2` | `(4096,18432,7168)` | 4.209 | 2.103 | 0.647 | 2.750 | 1.530x |
| `dsv3.ffn.w3` | `(4096,7168,18432)` | 4.209 | 2.103 | 0.647 | 2.750 | 1.530x |

Measured benchmark:

| layer | shape `(M,K,N)` | BF16 measured ms | FP8 measured ms | measured speedup | roofline speedup | speedup ratio of roofline |
|---|---:|---:|---:|---:|---:|---:|
| `dsv3.ffn.w1` | `(4096,7168,18432)` | 4.419 | 3.393 | 1.302x | 1.530x | 0.851 |
| `dsv3.ffn.w2` | `(4096,18432,7168)` | 4.233 | 3.271 | 1.294x | 1.530x | 0.845 |
| `dsv3.ffn.w3` | `(4096,7168,18432)` | 4.340 | 3.648 | 1.190x | 1.530x | 0.777 |
| average | | 4.330 | 3.437 | 1.262x | 1.530x | 0.825 |

The blockwise overhead model matches the current `Float8BlockwiseLinear`
implementation: input, weight, grad_output, transposed grad_output, and input
RHS layouts are modeled as separate quantization kernels, with weight
quantization modeled as two kernels for the transposed RHS layout used in
forward and the regular RHS layout used in `grad_input`.

The FP8 GEMM roofline intentionally ignores scale reads inside the GEMMs. For
these DSV3 shapes, scale reads are small relative to FP8 operand reads and BF16
output writes. Scale writes remain part of the quantization-overhead model.

### CODE REVIEW

**Comment by danielvegamyhre** [comment]
@iamzainhuda nice progress, i haven't looked at the code yet (lmk when it's ready for review) but please be sure to include the benchmark data for DSV3 671b shapes in the PR description when done

**Comment by iamzainhuda** [comment]
> @iamzainhuda nice progress, i haven't looked at the code yet (lmk when it's ready for review) but please be sure to include the benchmark data for DSV3 671b shapes in the PR description when done

added the benchmark results for both compile and non compiled runs with scaled_mm backend. i have another PR with a kernel optimization that should improve the speedup a bit

**Comment by danielvegamyhre** [comment]
@iamzainhuda please fix the ruff/linter issue

**Comment by iamzainhuda** [comment]
> @iamzainhuda please fix the ruff/linter issue

yup fixed

File: benchmarks/float8/float8_roofline.py [hunk 1]
```diff
@@ -238,6 +454,13 @@ def run(
     * `n_limit (optional)`: if specified, only runs `n_limit` iterations
     * `mx_recipe_name (optional)`: MX format recipe
     * `enable_fusion_modeling`: if False uses Linear, if True uses LNLinearSigmoid and models the fusion of float8 overhead
+    * `mx_recipe_name=blockwise_fp8_training`: benchmark the prototype
```
Comment (vkuzo) [inline benchmarks/float8/float8_roofline.py]: this is pretty hacky, IMO good options are:
1. just have your own hacky script (I did push back on this as rooflines are important and good to have some standardization)
2. follow the style in the current script

or, trade off somewhere between 1 and 2 and branch appropriately.  I'd say for the things you do want to reuse, good to follow a little bit better the style of the existing code.  Claude is good at this, just need to prompt.

File: benchmarks/float8/float8_roofline.py [hunk 2]
```diff
@@ -220,6 +243,194 @@ def do_matmul(A, B):
...
+    if shape_gen_name in ("dsv3-16b-671b", "dsv3-16b/671b", "dsv3_16b_671b"):
+        return get_dsv3_16b_671b_shapes_iter(
+            seq_len=dsv3_seq_len,
+            dim=dsv3_dim,
+            inter_dim=dsv3_inter_dim,
+        )
+    return get_name_to_shapes_iter(shape_gen_name, None, None, None)
+
+
+def get_blockwise_mem_time_sympy(bytes_rw, gpu_name: Optional[str] = None):
```
Comment (vkuzo) [inline benchmarks/float8/float8_roofline.py]: i'd put these next to the other ones (in `roofline_utils`) and make it consistent

File: benchmarks/float8/float8_roofline.py [hunk 3]
```diff
@@ -220,6 +243,194 @@ def do_matmul(A, B):
...
+
+    The memory model counts FP8 matrix reads, BF16 output writes, and FP32
+    scale reads. The GEMM estimate is max(compute time, memory time).
+    """
+    specs = get_specs(gpu_name)
+    gemm_ops = 2 * M * K * N
+    compute_gemm_time_s = (
+        gemm_ops / specs["fp8_peak_tops"] / specs["pct_achievable_gemm_tops"]
+    )
+    bytes_rw = (
```
Comment (vkuzo) [inline benchmarks/float8/float8_roofline.py]: add comments here to explain the kernel pattern you expect, the assumtions you are making, the terms you are ignoring, etc

File: benchmarks/float8/float8_roofline.py [hunk 4]
```diff
@@ -220,6 +243,194 @@ def do_matmul(A, B):
...
+
+
+def get_blockwise_quant_ovhd_s(
+    numel,
+    scale_numel,
+    gpu_name: Optional[str] = None,
+    fuse_read: bool = False,
+):
+    read_bytes = 0 if fuse_read else BYTES_PER_EL_BF16 * numel
+    bytes_rw = (
```
Comment (vkuzo) [inline benchmarks/float8/float8_roofline.py]: add some comments describing what your kernel is doing

File: benchmarks/float8/float8_roofline.py [hunk 5]
```diff
@@ -341,7 +611,7 @@ def run(
         rb_bf16_gemm_ratio = -1
         rb_fp8_gemm_ratio = -1
 
-        if do_benchmarks:
+        if do_benchmarks and not is_blockwise_fp8_training:
```
Comment (vkuzo) [inline benchmarks/float8/float8_roofline.py]: IMO this is too spaghetti, just do

```python
if do_benchmarks:
    assert not is_blockwise_fp8_training, "unsupported"
```

File: benchmarks/float8/float8_roofline.py [hunk 6]
```diff
@@ -416,13 +688,40 @@ def run(
 
             # get the bf16 gpu kernel time
             torch._dynamo.reset()
-            m_bf16 = torch.compile(copy.deepcopy(m_orig))
+            m_bf16 = copy.deepcopy(m_orig)
+            if (not is_blockwise_fp8_training) or blockwise_compile_benchmarks:
```
Comment (vkuzo) [inline benchmarks/float8/float8_roofline.py]: raise an exception on unsupported paths instead of this style

File: benchmarks/float8/float8_roofline.py [hunk 7]
```diff
@@ -220,6 +243,194 @@ def do_matmul(A, B):
...
+    Each modeled kernel is also lower-bounded by KERNEL_LAUNCH_OVERHEAD_SEC.
+    """
+    specs = get_specs(gpu_name)
+    mem_time_s = (
+        bytes_rw / specs["peak_mem_bw_bytes_sec"] / specs["pct_achievable_mem_bw"]
+    )
+    return sympy.Max(mem_time_s, KERNEL_LAUNCH_OVERHEAD_SEC)
+
+
+def get_blockwise_individual_gemm_time_sympy(
```
Comment (vkuzo) [inline benchmarks/float8/float8_roofline.py]: is this meaningfully different from the other fp8 gemm rooflines? my guess would be the scale numel can be dropped here without changing much in the overall results

**Comment by vkuzo** [comment]
@claude review this PR

File: benchmarks/float8/float8_roofline.py (hunk 2 above)
Comment (iamzainhuda) [reply benchmarks/float8/float8_roofline.py]: yup, moved over to roofline_utils

File: benchmarks/float8/float8_roofline.py (hunk 1 above)
Comment (iamzainhuda) [reply benchmarks/float8/float8_roofline.py]: got it, moved the shape generation to flaot8/utils, roofline math to roofline_utils, and removed this blockwise benchmark path. tried to keep the script as unchanged as possible

File: benchmarks/float8/float8_roofline.py (hunk 7 above)
Comment (iamzainhuda) [reply benchmarks/float8/float8_roofline.py]: yeah dropped this entirely so scaled reads are not part of the blockwise GEMM model but we still have the scale writes modeled in the quantization overhead

File: benchmarks/float8/float8_roofline.py (hunk 1 above)
Comment (iamzainhuda) [reply benchmarks/float8/float8_roofline.py]: with the fp8 blockwise linear cleanly benchmarking in this script didn't seem feasible so i split it off into a separate script. in this script we get the modeled roofline speedup now. the benchmark script uses this to get % of roofline speedup

File: benchmarks/float8/float8_roofline.py [hunk 8]
```diff
@@ -229,15 +238,24 @@ def run(
...
     Args:
     * `do_benchmarks`: if True, gemm and e2e fwd+bwd of LNLinearSigmoid are benchmarked
-    * `shape_gen_name`: `llama`, `pow2`, `pow2_extended`, or `sweep`
+    * `shape_gen_name`: `llama`, `pow2`, `pow2_extended`, `sweep`, or
+      `dsv3-16b-671b`
     * `gemm_cache_filename (optional)`: file to cache gemm benchmark results
     * `n_limit (optional)`: if specified, only runs `n_limit` iterations
     * `mx_recipe_name (optional)`: MX format recipe
     * `enable_fusion_modeling`: if False uses Linear, if True uses LNLinearSigmoid and models the fusion of float8 overhead
+    * `mx_recipe_name=blockwise_fp8_training`: use prototype
```
Comment (vkuzo) [inline benchmarks/float8/float8_roofline.py]: these aren't arguments, move to a section below?

File: benchmarks/float8/float8_roofline.py [hunk 9]
```diff
@@ -264,34 +287,66 @@ def run(
         "mxfp8_32x32_weight",
         # real mxfp4_cutlass recipe
         "mxfp4_cutlass",
+        # prototype DeepSeek-style blockwise FP8 training linear
+        BLOCKWISE_FP8_TRAINING_RECIPE_NAME,
```
Comment (vkuzo) [inline benchmarks/float8/float8_roofline.py]: why not `float8_recipe_name`? This is a float8 recipe.

File: benchmarks/float8/float8_roofline.py [hunk 10]
```diff
@@ -264,34 +287,66 @@ def run(
...
-    lowp_input_dtype = torch.float8_e4m3fn
-    if mx_recipe_name == "mxfp4_cutlass":
-        lowp_input_dtype = torch.float4_e2m1fn_x2
-
-    fp8_gemm_time_sympy = get_gemm_time_sympy(
-        M, K, N, lowp_input_dtype, float8_recipe_name, mx_recipe_name, None
+        M, K, N, torch.bfloat16, None, None, roofline_gpu_name
     )
+    if is_blockwise_fp8_training:
+        fp8_ovhd_time_sympy = get_blockwise_float8_mem_sympy(
```
Comment (vkuzo) [inline benchmarks/float8/float8_roofline.py]: `get_float8_mem_sympy` takes the recipe names, why not just be consistent if you are trying to reuse the code?

File: benchmarks/float8/float8_roofline.py [hunk 11]
```diff
@@ -264,34 +287,66 @@ def run(
...
+            M,
+            K,
+            N,
+            enable_fusion_modeling,
+            roofline_gpu_name,
+        )
+        fp8_gemm_time_sympy = get_blockwise_gemm_time_sympy(
```
Comment (vkuzo) [inline benchmarks/float8/float8_roofline.py]: same, `get_gemm_time_sympy` takes a recipe name

File: benchmarks/float8/float8_roofline.py [hunk 12]
```diff
@@ -446,11 +513,16 @@ def run(
         # Calculate e2e speedup if benchmarks were run, otherwise -1
         if b_bf16_e2e_time_s > 0 and b_fp8_e2e_time_s > 0:
             b_fp8_e2e_speedup = b_bf16_e2e_time_s / b_fp8_e2e_time_s
+            b_fp8_e2e_speedup_pct_of_r = (
+                b_fp8_e2e_speedup / r_fp8_gemm_and_ovhd_speedup * 100
```
Comment (vkuzo) [inline benchmarks/float8/float8_roofline.py]: iirc all the other ratios in this file do not multiply by 100 (i.e. 0.53 instead of 53%), IMO better to keep the units consistent

File: benchmarks/prototype/blockwise_fp8_training/bench_linear_roofline.py [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+
+    df = pd.DataFrame(rows)
+    pd.set_option("display.precision", 3)
+    print(df)
+    if args.outfile:
+        df.to_csv(args.outfile, index=False)
+        print(f"wrote {args.outfile}")
+
+
+def main():
```
Comment (vkuzo) [inline benchmarks/prototype/blockwise_fp8_training/bench_linear_roofline.py:249]: if we have this script customized for your use case, why do you need to change `benchmarks/float8/float8_roofline.py`?  i might be missing something

File: benchmarks/float8/float8_roofline.py (hunk 8 above)
Comment (iamzainhuda) [reply benchmarks/float8/float8_roofline.py]: agreed, i removed this entirely and kept these knobs in the benchmark script where it is CLI args

File: benchmarks/float8/float8_roofline.py (hunk 10 above)
Comment (iamzainhuda) [reply benchmarks/float8/float8_roofline.py]: yeah in my refactoring i missed this, benchmark script now calls the roofline helper directly

File: benchmarks/float8/float8_roofline.py (hunk 9 above)
Comment (iamzainhuda) [reply benchmarks/float8/float8_roofline.py]: yup, removed the blockwise mode from the shared script

File: benchmarks/float8/float8_roofline.py (hunk 11 above)
Comment (iamzainhuda) [reply benchmarks/float8/float8_roofline.py]: yup same as above comment

File: benchmarks/prototype/blockwise_fp8_training/bench_linear_roofline.py (hunk 13 above)
Comment (iamzainhuda) [reply benchmarks/prototype/blockwise_fp8_training/bench_linear_roofline.py:249]: yeah i agree, i don't actually need to change the shared script now. the added logic for fp88blockwise lives in the benchmark script instead now

**Review by vkuzo — APPROVED** [review]
lg!
